code/fmri_utils/func_preproc/optimize_map_coordinates.py

```diff
@@ -0,0 +1,154 @@
+# - compatibility with Python 2
+from __future__ import print_function  # print('me') instead of print 'me'
+from __future__ import division  # 1/2 == 0.5, not 0
+# - import common modules
+import numpy as np  # the Python array package
+import numpy.linalg as npl
+import matplotlib.pyplot as plt  # the Python plotting package
+import scipy.ndimage as snd
+from scipy.optimize import fmin_powell
+import nibabel as nib
+
+# import of rotations code
+#from rotations import x_rotmat, y_rotmat, z_rotmat
+from .rotations import x_rotmat, y_rotmat, z_rotmat
+
+# Mismatch function between two volumes
+def correl_mismatch(ref_vol, vol1):
+    """ Negative correlation between the two images, flattened to 1D
+
+    Parameters
+    ----------
+    ref_vol: array shape (I, J, K)
+        array with data from the reference volume
+
+    vol1: array shape (I, J, K)
+        array with data from the volume being transformed to the reference
+
+    Returns
+    -------
+    correl: float
+        correlation coefficient between the reference and transformed volumes
+
+    """
+    correl = np.corrcoef(ref_vol.ravel(), vol1.ravel())[0, 1]
+    return -correl
+
+# Resampling function for any given volume and the reference (ref_vol)
+def apply_coord_mapping(RB_params, ref_vol, vol1, ref_vol_affine):
+    """ Resample `ref_vol` to 'vol1' after applying a moving affine based on RB_params
+
+    Parameters
+    ----------
+    RB_params: array shape (6,)
+        array with rigid body transfomations (3 translations, 3 rotations) to be applied
+
+    ref_vol: array shape (I, J, K)
+        array with data from the reference volume
+
+    vol1: array shape (I, J, K)
+        array with data from the volume being transformed to the reference
+
+    ref_vol_affine: array shape (4, 4)
+        image affine mapping for the reference volume
+
+    Returns
+    -------
+    resampled_vol1: arra shape (I, J, K)
+        vol1 resampled according to the rigid body parameters
+
+    """
+    # get rotation value sin each direction and make a rotation matrix
+    r_x, r_y, r_z = RB_params[3:]
+    rot_mat = z_rotmat(r_z).dot(y_rotmat(r_y).dot(x_rotmat(r_x)))
+    # get translation values in each direciton
+    trans = np.array(RB_params[0:3])
+    moving_affine = nib.affines.from_matvec(rot_mat, trans)
+
+    # affine maping bet the moving affine and the affine of the reference volume
+    vol1_to_ref_vol = npl.inv(moving_affine).dot(ref_vol_affine)
+
+    # get shape of reference volume
+    I, J, K = ref_vol.shape
+    # create a mesh coordinate grid
+    i_vals, j_vals, k_vals = np.meshgrid(range(I), range(J), range(K), indexing='ij')
+    in_vox_coords = np.array([i_vals, j_vals, k_vals])
+
+    #create array of 3*(I,J,K) random numbers between -0.5 and 0.5
+    jitter = np.random.uniform(-0.5,0.5, 3*I*J*K)
```

Comment:
Or `jitter = np.random.uniform(-0.5,0.5, size=in_vox_coords.shape)`.

Why not try:

* turning down the jitter (-0.1 to 0.1)
* using a single jitter to realign all the volumes in the run?